File: tools/script/compile.ts
#!/usr/bin/env ts-node

// Date : 2022-07-01
// A script to compile Library cgraph in Unix-like and Windows Platforms
// gets source files iteratively from Directory src

const syspath = require("path");
const fs = require("fs");
const os = require("os");
const child_process = require("child_process");

let PRO: string = "cgraph";
let DIR: string = ".";
let INC: string = syspath.join(DIR, "include");
let SRC: string = syspath.join(DIR, "src");
let SRC_TYPE: string = syspath.join(SRC, "type");
let TST: string = syspath.join(DIR, "tests");
let LIB: string = syspath.join(DIR, "lib");

let CC: string = "cc";
let CFLAGS: string = "-std=c89 -Wall -pedantic -fPIC";
let CSFLAGS: string = "-shared";

let MODE: string = "debug";
if (MODE == "debug") {
  CFLAGS.concat(" -g -DDEBUG");
} else if (MODE == "release") {
  CFLAGS.concat(" -static -O2");
}

// package shared library
let AR: string = "ar";
let ARFLAGS: string = "-rcs";

// source files
// get all subdirectories
function getsubdirs(path: string, dirs: string[]) {
  const files: string[] = fs.readdirSync(path);
  for (let i: number = 0; i < files.length; i++) {
    let subpath: string  = syspath.join(path, files[i]);
    let stats: any = fs.lstatSync(subpath);
    if (stats.isDirectory() && !files[i].startsWith(".")) {
      dirs.push(subpath);
      getsubdirs(subpath, dirs);
    }
  }
}

// get all source files from subdirectories
let SRCS: string[] = [];
getsubdirs(SRC, SRCS);

let CFILES: string[] = [];
for (let i: number = 0; i < SRCS.length; i++) {
  const files: string[] = fs.readdirSync(SRCS[i]);
  for (let j: number = 0; j < files.length; j++) {
    let file: string = syspath.join(SRCS[i], files[j]);
    let stats: any = fs.lstatSync(file);
    if (stats.isFile() && /^((?!\.).)*\.c$/.test(files[j])) {
      CFILES.push(file);
    }
  }
}

let LIBSHARED: string;
let LIBSTATIC: string;
let TSTSUFFIX: string;
if ("Windows_NT" == os.type()) {
  // target files
  LIBSHARED = syspath.join(LIB, `lib${PRO}.dll`);
  LIBSTATIC = syspath.join(LIB, `lib${PRO}.lib`);
  // test files
  TSTSUFFIX = ".exe";
} else {
  // target files
  LIBSHARED = syspath.join(LIB, `lib${PRO}.so`);
  LIBSTATIC = syspath.join(LIB, `lib${PRO}.a`);
  // test files
  TSTSUFFIX = "";
}

if (process.argv.length == 2) {
  if (!fs.existsSync(LIB)) {
    fs.mkdirSync(LIB);
  }
  let OFILES = [];
  for (let i: number = 0; i < CFILES.length; i++) {
    let obj: string = CFILES[i].replace(/\.c$/, ".o");
    let dep: string = CFILES[i].replace(/\.c$/, ".d");
    console.log(`compile ${CFILES[i]} to ${obj}`);
    child_process.exec(`${CC} ${CFLAGS} -I${INC} -I${SRC_TYPE} -c ${CFILES[i]} -o ${obj} -MD -MF ${dep}`);
    OFILES.push(obj);
  }
  console.log(`compile ${LIBSHARED}`);
  child_process.exec(`${CC} ${CSFLAGS} -o ${LIBSHARED} ${OFILES.join(" ")}`);
  console.log(`compile ${LIBSTATIC}`);
  child_process.exec(`${AR} ${ARFLAGS} ${LIBSTATIC} ${OFILES.join(" ")}`);
} else if (process.argv[2] == "test") {
  const files: string[] = fs.readdirSync(TST);
  for (let i: number = 0; i < files.length; i++) {
    let file: string = syspath.join(TST, files[i]);
    let stats: any = fs.lstatSync(file);
    if (stats.isFile() && /^((?!\.).)*\.c$/.test(files[i])) {
      let target: string = file.replace(/\.c$/, TSTSUFFIX);
      console.log(`compile ${file} to ${target}`);
      child_process.exec(`${CC} ${CFLAGS} -I${INC} -o ${target} ${file} -L${LIB} -static -l${PRO} -lm`);
      console.log(`test ${target}`);
      child_process.exec(target);
    }
  }
} else if (process.argv[2] == "clean") {
  for (let i: number = 0; i < CFILES.length; i++) {
    let obj: string = CFILES[i].replace(/\.c$/, ".o");
    console.log(`clean ${obj}`);
    if (fs.existsSync(obj)) {
      fs.unlink(obj);
    }
    let dep: string = CFILES[i].replace(/\.c$/, ".d");
    console.log(`clean ${dep}`);
    if (fs.existsSync(dep)) {
      fs.unlink(dep);
    }
  }
  console.log(`clean ${LIBSTATIC}`);
  if (fs.existsSync(LIBSTATIC)) {
    fs.unlink(LIBSTATIC);
  }
  console.log(`clean ${LIBSHARED}`);
  if (fs.existsSync(LIBSHARED)) {
    fs.unlink(LIBSHARED);
  }
} else if (process.argv[2] == "distclean") {
  for (let i: number = 0; i < CFILES.length; i++) {
    let obj: string = CFILES[i].replace(/\.c$/, ".o");
    console.log(`clean ${obj}`);
    if (fs.existsSync(obj)) {
      fs.unlink(obj);
    }
    let dep: string = CFILES[i].replace(/\.c$/, ".d");
    console.log(`clean ${dep}`);
    if (fs.existsSync(dep)) {
      fs.unlink(dep);
    }
  }
  console.log(`clean ${LIBSTATIC}`);
  if (fs.existsSync(LIBSTATIC)) {
    fs.unlink(LIBSTATIC);
  }
  console.log(`clean ${LIBSHARED}`);
  if (fs.existsSync(LIBSHARED)) {
    fs.unlink(LIBSHARED);
  }
  console.log(`clean ${LIB}`);
  if (fs.existsSync(LIB)) {
    fs.rmSync(LIB);
  }
} else if (process.argv[2] == "help") {
  console.log(`${process.argv[1]} <target>`);
  console.log("<target>: ");
  console.log("                    compile cgraph");
  console.log("          test      test cgraph");
  console.log("          clean     clean all the generated files");
  console.log("          distclean clean all the generated files and directories");
  console.log("          help      commands to this program");
} else {
  console.log(`${process.argv[2]} is an unsupported command`);
  console.log(`use \"${process.argv[1]} help\" to know all supported commands`);
}

File: tools/script/compile.java
//!/usr/bin/java --source 11

// Date : 2022-07-01
// A script to compile Library cgraph in Unix-like and Windows Platforms
// gets source files iteratively from Directory src

import java.util.List;
import java.util.ArrayList;
import java.io.File;
import java.lang.Runtime;

public class Compile {
  public static void getsubdirs(String path, List<String> dirs) {
    File[] items = new File(path).listFiles()
    for (File item : items) {
      String filename = item.getName();
      if (item.isDir() && !filename.startsWith(".")) {
        dirs.add(item.getPath());
        getsubdirs(item.getPath(), dirs);
      }
    }
  }

  public static void main(String[] args) {
    String PRO = new String("cgraph");
    String DIR = new String(".");
    String INC = new String(DIR + File.separator + "include");
    String SRC = new String(DIR + File.separator + "src");
    String TST = new String(DIR + File.separator + "test");
    String LIB = new String(DIR + File.separator + "lib");

    String CC = new String("cc");
    StringBuffer CFLAGS = new StringBuffer("-std=c89 -Wall -pedantic -fPIC");
    String CSFLAGS = new String("-shared");

    String MODE = new String("debug");
    if (MODE.equals("debug")) {
      CFLAGS.append(" -g -DDEBUG");
    } else if (MODE.equals("release")) {
      CFLAGS.append(" -static -O2");
    }

    // package shared library
    String AR = new String("ar");
    String ARFLAGS = new String("-rcs");
    
    String OS = System.getProperty("os.name");
    String LIBSHARED, LIBSTATIC, TSTFILE, TSTTARGET;
    if (OS.matches("^[Ww]in")) {
      // target files
      LIBSHARED = new String(LIB + File.separator + "lib" + PRO + ".dll");
      LIBSTATIC = new String(LIB + File.separator + "lib" + PRO + ".lib");
      // test files
      TSTFILE = new String(TST + File.separator + PRO + ".c");
      TSTTARGET = new String(TST + File.separator + PRO + ".exe");
    } else {
      // target files
      LIBSHARED = new String(LIB + File.separator + "lib" + PRO + ".so");
      LIBSTATIC = new String(LIB + File.separator + "lib" + PRO + ".a");
      // test files
      TSTFILE = new String(TST + File.separator + PRO + ".c");
      TSTTARGET = new String(TST + File.separator + PRO);
    }
    
    // source files
    // get all subdirectories
    List<String> SRCS = new ArrayList<String>();
    getsubdirs(SRC, SRCS);

    // get all source files from subdirectories
    List<File> CFILES = new ArrayList<File>();
    for (String dir : SRCS) {
      File[] FILES = new File(dir).listFiles();
      for (File file : FILES) {
        String filename = file.getName();
        if (file.isFile() && !filename.startsWith(".") && filename.endsWith(".c")) {
          CFILES.add(file);
        }
      }
    }


    String SCRIPT_NAME = Thread.currentThread().getStackTrace()[1].getFileName();
    if (args.length == 0) {
      if (!(new File(LIB).exists())) {
        new File(LIB).mkdir();
      }
      List<String> OFILES = new ArrayList<String>();
      for (File file : CFILES) {
        String obj = file.getPath().replace(".c", ".o");
        try {
          String dep = file.getPath().replace(".c", ".d");
          String cmd = String.format("%s %s -I%s -c %s -o %s -MD -MF %s", CC, CFLAGS, INC, file, obj, dep);
          System.out.println(String.format("compile %s to %s", file, obj));
          Process program = Runtime.getRuntime().exec(cmd);
          OFILES.add(obj);
        } catch (Exception e) {
          System.out.println(String.format("ERROR: compile %s to %s", file, obj));
        }
      }
      try {
        String cmd = String.format("%s %s -o %s %s", CC, CSFLAGS, LIBSHARED, String.join(" ", OFILES));
        System.out.println(String.format("compile %s", LIBSHARED));
        Process program = Runtime.getRuntime().exec(cmd);
      } catch (Exception e) {
        System.out.println(String.format("ERROR: compile *.o to %s", LIBSHARED));
      }
      try {
        String cmd = String.format("%s %s %s %s", AR, ARFLAGS, LIBSTATIC, String.join(" ", OFILES));
        System.out.println(String.format("compile %s", LIBSTATIC));
        Process program = Runtime.getRuntime().exec(cmd);
      } catch (Exception e) {
        System.out.println(String.format("ERROR: compile *.o to %s", LIBSTATIC));
      }
    } else if (args[0].equals("test")) {
      try {
        String cmd = String.format("%s %s -I%s -o %s %s -L%s -static -l%s -lm", CC, CFLAGS, INC, TSTTARGET, TSTFILE, LIB, PRO);
        System.out.println(String.format("compile %s to %s", TSTFILE, TSTTARGET));
        Process program = Runtime.getRuntime().exec(cmd);
      } catch (Exception e) {
        System.out.println(String.format("ERROR: compile %s to %s", TSTFILE, TSTTARGET));
      }
      try {
        String cmd = String.format("%s %s", TSTTARGET, TST + File.separator + "elements.csv");
        System.out.println(String.format("test %s with %s", TSTTARGET, TST + File.separator + "elements.csv"));
        Process program = Runtime.getRuntime().exec(cmd);
      } catch (Exception e) {
        System.out.println(String.format("ERROR: test %s with %s", TSTTARGET, TST + File.separator + "elements.csv"));
      }
    } else if (args[0].equals("clean")) {
      for (File file : CFILES) {
        File obj = new File(file.getPath().replace(".c", ".o"));
        System.out.println(String.format("clean %s", obj));
        obj.delete();
        File dep = new File(file.getPath().replace(".c", ".d"));
        System.out.println(String.format("clean %s", dep));
        dep.delete();
      }
      System.out.println(String.format("clean %s", LIBSTATIC));
      File libstatic = new File(LIBSTATIC);
      libstatic.delete();
      System.out.println(String.format("clean %s", LIBSHARED));
      File libshared = new File(LIBSHARED);
      libshared.delete();
      System.out.println(String.format("clean %s", TSTTARGET));
      File tsttarget = new File(TSTTARGET);
      tsttarget.delete();
    } else if (args[0].equals("distclean")) {
      for (File file : CFILES) {
        File obj = new File(file.getPath().replace(".c", ".o"));
        System.out.println(String.format("clean %s", obj));
        obj.delete();
        File dep = new File(file.getPath().replace(".c", ".d"));
        System.out.println(String.format("clean %s", dep));
        dep.delete();
      }
      System.out.println(String.format("clean %s", LIBSTATIC));
      File libstatic = new File(LIBSTATIC);
      libstatic.delete();
      System.out.println(String.format("clean %s", LIBSHARED));
      File libshared = new File(LIBSHARED);
      libshared.delete();
      System.out.println(String.format("clean %s", LIB));
      File lib = new File(LIB);
      lib.delete();
      System.out.println(String.format("clean %s", TSTTARGET));
      File tsttarget = new File(TSTTARGET);
      tsttarget.delete();
    } else if (args[0].equals("help")) {
      System.out.println(SCRIPT_NAME +" <target>");
      System.out.println("<target>: ");
      System.out.println("                    compile cgraph");
      System.out.println("          test      test cgraph");
      System.out.println("          clean     clean all the generated files");
      System.out.println("          distclean clean all the generated files and directories");
      System.out.println("          help      commands to this program");
    } else {
      System.out.println(args[0] + " is an unsupported command");
      System.out.println("use \"" + SCRIPT_NAME + " help\" to know all supported commands");
    }
  }
}